Limit max batch size for segment allocation, add docs

kfaraz opened the pull request:
Follow up to #13369 

Changes:
- Limit max batch size in `SegmentAllocationQueue` to 500
- Rename `batchAllocationMaxWaitTime` to `batchAllocationWaitTime` since it was a little misleading and actual wait time may exceed this configured value.
- Minor refactor: remove usage of `SegmentInsertAction` from `TaskToolbox`

gianm commented on docs/ingestion/tasks.md:
```diff
@@ -343,6 +343,27 @@ You can override the task priority by setting your priority in the task context
+
+### Batching `segmentAllocate` actions
+
+In a cluster with a large number of concurrent tasks (say > 1000), `segmentAllocate` actions on the overlord may take very long intervals of time to finish thus causing spikes in the `task/action/run/time`. This may result in lag building up while a task waits for a segment to get allocated.
```
I think this is going to be useful even at much lower concurrent task counts. I have seen people have issues with allocation timings around task rollover even with just 10s of tasks, requiring a scale-up of metadata store size.

gianm commented on docs/ingestion/tasks.md:
```diff
@@ -343,6 +343,27 @@ You can override the task priority by setting your priority in the task context
+These are various overlord actions performed by tasks during their lifecycle. Some typical actions are as follows:
+- `lockAcquire`: acquires a time-chunk lock on an interval for the task 
+- `lockRelease`: releases a lock acquired by the task on an interval
+- `segmentInsertion`: inserts segments into metadata store
```
Worth mentioning `segmentTransactionalInsert` too?

gianm commented on docs/operations/metrics.md:
```diff
@@ -230,8 +230,14 @@ Note: If the JVM does not support CPU time measurement for the current thread, `
+|`task/action/log/time`|Milliseconds taken to log a task action to the audit log.| `dataSource`, `taskId`, `taskType`, `taskActionType`|< 1000 (subsecond)|
+|`task/action/run/time`|Milliseconds taken to execute a task action.| `dataSource`, `taskId`, `taskType`, `taskActionType`|Varies from subsecond to a few seconds, based on action type.|
+|`task/action/success/count`|Number of task actions that were executed successfully during the emission period. Currently only being emitted for batched `segmentAllocate` actions.| `dataSource`, `taskId`, `taskType`, `taskActionType`|Varies|
+|`task/action/failed/count`|Number of task actions that failed during the emission period. Currently only being emitted for batched `segmentAllocate` actions.| `dataSource`, `taskId`, `taskType`, `taskActionType`|Varies|
```
Would be good to link all mentions of "batched segmentAllocate actions" to `../ingestion/tasks.md#batching-segmentallocate-actions`.

gianm commented on docs/configuration/index.md:
```diff
@@ -1112,6 +1112,8 @@ These Overlord static configurations can be defined in the `overlord/runtime.pro
 |`druid.indexer.storage.recentlyFinishedThreshold`|Duration of time to store task results. Default is 24 hours. If you have hundreds of tasks running in a day, consider increasing this threshold.|PT24H|
 |`druid.indexer.tasklock.forceTimeChunkLock`|_**Setting this to false is still experimental**_<br/> If set, all tasks are enforced to use time chunk lock. If not set, each task automatically chooses a lock type to use. This configuration can be overwritten by setting `forceTimeChunkLock` in the [task context](../ingestion/tasks.md#context). See [Task Locking & Priority](../ingestion/tasks.md#cont ...
+|`druid.indexer.tasklock.batchSegmentAllocation`| If set to true, segment allocate actions are performed in batches to improve the throughput and reduce the average `task/action/run/time`.|false|
+|`druid.indexer.tasklock.batchAllocationWaitTime`|Milliseconds to wait between adding the first segment allocate action to a batch and executing that batch. The waiting time allows the batch to add more requests and thus improve the average segment allocation run time. This configuration takes effect only if `batchSegmentAllocation` is enabled. <br> This value should be decreased __only if__ ther ...
```
500 milliseconds already seems pretty short. If people are getting failures due to too-large batches, I wonder if reducing this will reliably fix their problem. Maybe it's better to suggest they turn off batch segment allocation completely.

Ideally, the batching would be robust to this case, and avoid generating oversized metadata operations. Is that possible?

kfaraz replied:
Yes, we can limit the batch size to a (hard-coded) max value, say 500. We have seen clusters where batches of 700-800 have executed successfully.

gianm commented on docs/configuration/index.md:
```diff
@@ -1112,6 +1112,8 @@ These Overlord static configurations can be defined in the `overlord/runtime.pro
 |`druid.indexer.storage.type`|Choices are "local" or "metadata". Indicates whether incoming tasks should be stored locally (in heap) or in metadata storage. "local" is mainly for internal testing while "metadata" is recommended in production because storing incoming tasks in metadata storage allows for tasks to be resumed if the Overlord should fail.|local|
 |`druid.indexer.storage.recentlyFinishedThreshold`|Duration of time to store task results. Default is 24 hours. If you have hundreds of tasks running in a day, consider increasing this threshold.|PT24H|
 |`druid.indexer.tasklock.forceTimeChunkLock`|_**Setting this to false is still experimental**_<br/> If set, all tasks are enforced to use time chunk lock. If not set, each task automatically chooses a lock type to use. This configuration can be overwritten by setting `forceTimeChunkLock` in the [task context](../ingestion/tasks.md#context). See [Task Locking & Priority](../ingestion/tasks.md#cont ...
+|`druid.indexer.tasklock.batchSegmentAllocation`| If set to true, segment allocate actions are performed in batches to improve the throughput and reduce the average `task/action/run/time`.|false|
```
Would be good to link to `../ingestion/tasks.md#batching-segmentallocate-actions`.

kfaraz commented:
Thanks a lot for the feedback, @gianm !

I have made the following changes:
- Limited max batch size to 500
- Updated links and docs based on feedback
- Added doc about `segmentTransactionalInsert` and removed doc of `segmentInsert`. 
- A `segmentInsert` action always translates to a `segmentTransactionInsert` action (see snippet below). It was being used only in the `TaskToolbox`. I have updated this usage.

I wonder if we should just deprecate and remove the `segmentInsert` action.
Please let me know if the changes make sense.

https://github.com/apache/druid/blob/136322d13b68f1e14552ea606ac70101ff6af400/indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentInsertAction.java#L71-L75

writer-jill commented on docs/ingestion/tasks.md:
```diff
@@ -343,6 +343,27 @@ You can override the task priority by setting your priority in the task context
+- `lockRelease`: releases a lock acquired by the task on an interval
+- `segmentTransactionalInsert`: publishes new segments created by a task and optionally overwrites and/or drops existing segments in a single transaction
+- `segmentAllocate`: allocates pending segments to a task to write rows
+- etc.
```
```suggestion
```

writer-jill commented on docs/ingestion/tasks.md:
```diff
@@ -343,6 +343,27 @@ You can override the task priority by setting your priority in the task context
+
+### Batching `segmentAllocate` actions
+
+In a cluster with several concurrent tasks, `segmentAllocate` actions on the overlord may take very long intervals of time to finish thus causing spikes in the `task/action/run/time`. This may result in ingestion lag building up while a task waits for a segment to get allocated.
```
```suggestion
In a cluster with several concurrent tasks, `segmentAllocate` actions on the overlord can take a long time to finish, causing spikes in the `task/action/run/time`. This can result in ingestion lag building up while a task waits for a segment to be allocated.
```

writer-jill commented on docs/ingestion/tasks.md:
```diff
@@ -343,6 +343,27 @@ You can override the task priority by setting your priority in the task context
+### Batching `segmentAllocate` actions
+
+In a cluster with several concurrent tasks, `segmentAllocate` actions on the overlord may take very long intervals of time to finish thus causing spikes in the `task/action/run/time`. This may result in ingestion lag building up while a task waits for a segment to get allocated.
+The root causes of such spikes are:
```
```suggestion
The root cause of such spikes is likely to be one of more of the following:
```

writer-jill commented on docs/ingestion/tasks.md:
```diff
@@ -343,6 +343,27 @@ You can override the task priority by setting your priority in the task context
+- large number of metadata calls made to the segments and pending segments tables 
+- concurrency limitations while acquiring a task lock required for allocating a segment
+
+Since the contention typically arises from tasks allocating segments for the same datasource and interval, the run times can be improved by batching the actions together.
```
```suggestion
Since the contention typically arises from tasks allocating segments for the same datasource and interval, you can improve the run times by batching the actions together.
```

writer-jill commented on docs/ingestion/tasks.md:
```diff
@@ -343,6 +343,27 @@ You can override the task priority by setting your priority in the task context
+- concurrency limitations while acquiring a task lock required for allocating a segment
+
+Since the contention typically arises from tasks allocating segments for the same datasource and interval, the run times can be improved by batching the actions together.
+Batched segment allocation can be enabled on the overlord by setting `druid.indexer.tasklock.batchSegmentAllocation=true`.See [overlord configuration](../configuration/index.md#overlord-operations) for more details.
```
```suggestion
To enable batched segment allocation on the overlord, set  `druid.indexer.tasklock.batchSegmentAllocation` to  `true`. See [overlord configuration](../configuration/index.md#overlord-operations) for more details.
```

writer-jill reviewed (commented):
Added a few suggestions for style & clarity.

gianm commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java:
```diff
@@ -198,6 +199,24 @@ public Future<SegmentIdWithShardSpec> add(SegmentAllocateRequest request)
+      }
+    }
+
+    throw new ISE("Allocation queue is at capacity, all batches are full.");
```
Is `MAX_QUEUE_SIZE` measured in batches or allocations? What happens if it gets hit? How likely is it that we will hit it?

kfaraz replied:
`MAX_QUEUE_SIZE` applies to batches. I am using it in the loop above because that is the maximum possible number of batches, even for the same datasource, group id, etc

It is extremely unlikely to hit the max queue limit. It can happen if say, we have 2000 datasources, all doing streaming ingestion concurrently. If the queue is full, new incoming allocate actions are rejected.

gianm replied:
Got it. Sounds good to me.

kfaraz commented:
Merging this as the failure is unrelated. The same test has passed for jdk11.

maytasm commented:
@kfaraz @gianm Any reason why we did not make batch size configurable? If we are hitting batch size limit then increasing batchAllocationWaitTime will not help us.
